### PR TITLE
React Native Clientlib MWA 2.0

### PR DESCRIPTION
**Pull request by Michaelsulistio**
**Description**

This PR aims to update the React Native MWA libraries to MWA 2.0 API. Updating example apps can happen in a separate PR.

Updated libraries:
- `@solana-mobile/mobile-wallet-adapter-protocol`
- `@solana-mobile/mobile-wallet-adapter-protocol-web3js`

#### `@solana-mobile/mobile-wallet-adapter-protocol`
- Update `/android/` dependencies in `build.gradle` to use `com.solanamobile:mobile-wallet-adapter-clientlib:2.0.0-alpha5`
- Add deprecation warnings to `signTransactions`, `reauthorize`, and legacy `authorize` API
- Add new MWA 2.0 parameters to `options` in `signAndSendTransactions` API
- `signMessages` deprecation + new `signMessagesDetached` API (**TODO**: revert these changes)

#### `@solana-mobile/mobile-wallet-adapter-protocol-web3js`
- Add deprecation warning to `signTransactions`
- Add new MWA 2.0 parameters to `options` in `signAndSendTransactions` API
- `signMessages` deprecation + new `signMessagesDetached` API (**TODO**: revert these changes)

### Test Plan
To test the PR, I made local changes to the `example-react-native-app` repo and manually ran test cases.

**Setup**
1. Temporarily change the `package.json` of `@solana-mobile/mobile-wallet-adapter-protocol-web3js` to:
```
    "dependencies": {
        "@solana-mobile/mobile-wallet-adapter-protocol": "file:../mobile-wallet-adapter-protocol",
        ....
    }
```
2. In `/mobile-wallet-adapter-protocol/`, run `yarn build` to rebuild exported `lib/`
3. In `/mobile-wallet-adapter-protocol-web3js/`, run `rm -rf node_modules` and `yarn install` to ensure clean install of `mobile-wallet-adapter-protocol`
4. Now you can run `example-react-native-app` with `npx react-native start`

### CODE REVIEW

File: js/packages/mobile-wallet-adapter-protocol-web3js/src/transact.ts [hunk 1]
```diff
@@ -111,6 +134,27 @@ export async function transact<TReturn>(
                                 return signedMessages;
                             } as Web3MobileWallet[TMethodName];
                             break;
+                        case 'signMessagesDetached':
```
Comment (Michaelsulistio) [inline js/packages/mobile-wallet-adapter-protocol-web3js/src/transact.ts:137]: ditto. Ignore/revert these changes as needed

File: js/packages/mobile-wallet-adapter-protocol/android/build.gradle [hunk 2]
```diff
@@ -132,7 +132,7 @@ def kotlin_version = getExtOrDefault('kotlinVersion')
 dependencies {
   //noinspection GradleDynamicVersion
   implementation "com.facebook.react:react-native:+"  // From node_modules
-  implementation "com.solanamobile:mobile-wallet-adapter-clientlib:1.2.0"
+  implementation "com.solanamobile:mobile-wallet-adapter-clientlib:2.0.0-alpha5"
```
Comment (Michaelsulistio) [inline js/packages/mobile-wallet-adapter-protocol/android/build.gradle:135]: Updating the underlying native module to use the new clientlib with MWA 2.0 changes

File: js/packages/mobile-wallet-adapter-protocol/android/gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -1,5 +1,6 @@
+#Tue Nov 21 15:17:20 PST 2023
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.5-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.0-bin.zip
```
Comment (Michaelsulistio) [inline js/packages/mobile-wallet-adapter-protocol/android/gradle/wrapper/gradle-wrapper.properties:4]: Android Studio prompted me to upgrade this in order to build it successfully. Is this necessary?

File: js/packages/mobile-wallet-adapter-protocol/src/types.ts [hunk 4]
```diff
@@ -80,12 +99,30 @@ export interface GetCapabilitiesAPI {
 export interface ReauthorizeAPI {
     reauthorize(params: { auth_token: AuthToken; identity: AppIdentity }): Promise<AuthorizationResult>;
```
Comment (Michaelsulistio) [inline js/packages/mobile-wallet-adapter-protocol/src/types.ts:100]: Do we want a deprecation warning on `reauthorize`? I noticed that clientlib-ktx [doesn't have one.](https://github.com/solana-mobile/mobile-wallet-adapter/blob/main/android/clientlib-ktx/src/main/java/com/solana/mobilewalletadapter/clientlib/LocalAdapterOperations.kt#L48C1-L49C1)

**Comment by Funkatronics** [comment]
superseded by #623
